### DISCUSSION TITLE
Populating commands from config object?

### DISCUSSION THREAD

**Discussion by pawamoy** [DanCardin/cappa#177] [Q&A] [answered]
How would you go about populating a command's attributes from a config object :thinking:?

Here's what I'm playing with currently:

```python
# --------------------------------------------
# Declare the config stuff
# --------------------------------------------
class Unset:
    """A sentinel value for unset configuration options."""

    def __repr__(self) -> str:
        return "<unset>"


@dataclass(kw_only=True)
class Config:
    """Configuration for the insiders project."""

    unset: ClassVar[Unset] = Unset()

    sponsors_minimum_amount: int | Unset = unset

    # ...more config fields


@cappa.command(name="insiders")
@dataclass(kw_only=True)
class CommandMain:
    """Command to manage your Insiders projects."""

    subcommand: An[
        CommandBacklog,
        cappa.Subcommand(group=_GROUP_SUBCOMMANDS),
        Doc("The selected subcommand."),
    ]

    # --------------------------------------------
    # Main command propagates a config arg
    # --------------------------------------------
    config: An[
        Config | None,
        cappa.Arg(
            short="-c",
            long=True,
            parse=Config.from_file,
            default=Config.from_default_location(),
            group=_GROUP_GLOBAL,
            propagate=True,
        ),
        Doc("Path to the configuration file."),
    ] = None


@cappa.command(name="backlog")
@dataclass(kw_only=True)
class CommandBacklog:
    """Command to list the backlog of issues."""

    github_namespaces: An[
        list[str],
        cappa.Arg(num_args=-1),
        Doc("Namespaces to fetch issues from."),
    ]

    github_org_users: An[
        dict[str, list[str]],
        cappa.Arg(short=False, long=True),
        Doc("""A mapping of users belonging to sponsoring organizations."""),
    ]

    github_token: An[
        str,
        cappa.Arg(short=False, long=True, default=cappa.Env("GITHUB_TOKEN")),
        Doc("""A GitHub token. Recommended scopes: `repo`."""),
    ]

    polar_token: An[
        str,
        cappa.Arg(short=False, long=True, default=cappa.Env("POLAR_TOKEN")),
        Doc("""A Polar token. Recommended scopes: `read:backlog`."""),
    ]

    # ------------------------------------------------------------
    # Subcommands depend on main to fetch config values
    # ------------------------------------------------------------
    def __call__(self, main: CommandMain) -> int:
        if main.config:
            if not self.github_namespaces and main.config.backlog_namespaces is not Config.unset:
                self.github_namespaces = main.config.backlog_namespaces
            if not self.github_org_users and main.config.github_organization_members is not Config.unset:
                self.github_org_users = main.config.github_organization_members
            if not self.sort and main.config.backlog_sort is not Config.unset:
                self.sort = main.config.backlog_sort
            if not self.github_token and main.config.github_token_command is not Config.unset:
                self.github_token = main.config.github_token_command
            if not self.polar_token and main.config.polar_token_command is not Config.unset:
                self.polar_token = main.config.polar_token_command

        # ...actual code
```

I don't think this approach is the best, because required attributes like `github_token` will emit an error *before* they can be populated from the config, if they aren't provided from the CLI or through an env var.

Is there a more Cappa-idiomatic way to declare such an interface? The command dataclasses should keep representing the *final state of the data*, while each attribute is *annotated to tell how it is populated*.

I'm thinking about [Pydantic settings](https://docs.pydantic.dev/latest/concepts/pydantic_settings/#usage), wondering if there's a way to integrate it with Cappa somehow?

**Comment by pawamoy** [comment]
I'm imagining something like `cappa.Env`.

Here we pass a tuple to `cappa.From` (dumb name for the example), which is the type of the command to be injected, and an attribute path, to get and return:

```python
    github_token: An[
        str,
        cappa.Arg(
            short=False,
            long=True,
            default=[
                cappa.Env("GITHUB_TOKEN"),
                cappa.From(Command, "config.github_token"),
            ],
        ),
        Doc("A GitHub token. Recommended scopes: `repo`."),
    ]
```

Here we pass it a callable directly, which expects a `Command` instance again:

```python
from functools import partial

unset = ...  # sentinel

def from_config(main: Command, attribute_name: str):
    if main.config and (attr := getattr(main.config, attribute_name, unset)) is not unset:
        return attr
    raise ValueError("whatever, just to instruct Cappa that there was no value?")


# ...in our dataclass:

    github_token: An[
        str,
        cappa.Arg(
            short=False,
            long=True,
            default=[
                cappa.Env("GITHUB_TOKEN"),
                cappa.From(partial(from_config, attribute_name="github_token")),
            ],
        ),
        Doc("A GitHub token. Recommended scopes: `repo`."),
    ]
```

:thinking: :thinking: :thinking:

**Reply by pawamoy** [reply]
If accepting a `list` as `default` is too disruptive, could imagine supporting `or` / `|` :

```python
default=cappa.Env("GITHUB_TOKEN") or cappa.From(Command, "config.github_token")
default=cappa.Env("GITHUB_TOKEN") | cappa.From(Command, "config.github_token")
```

**Reply by DanCardin** [reply]
> I'm thinking about [Pydantic settings](https://docs.pydantic.dev/latest/concepts/pydantic_settings/#usage), wondering if there's a way to integrate it with Cappa somehow?

Well i'd firstly suggest https://github.com/DanCardin/dataclass-settings instead of dataclass-settings 😆

---

I think injecting from a `Command` instance is problematic because you're saying you need an instance of `Command` to already be instantiated in order for the command to then be updated with new overriding values.

Or i guess perhaps in at least this specific case, (in theory) a Command **could** be instantiated beforehand because you're only making use of it inside the context of a subcommand class. I **do** think the feasibility of **this** avenue very much depends on it being only accessible from a subcommand.

My thought being that perhaps you could supply a `parse` (or a From is syntactic construct for a `parse`) which could "depend" on a command `From(from_config)` `def from_config(main: Command):`. Which would leave the specific method by which you pull out a specific field outside cappa since, I'm less certain I love this `cappa.From(Command, "config.github_token")`.

---

Another thought that occurs to me, (which might not be ideal because it requires every field of `Config` to be optional) is that perhaps you could do something like

```
@dataclass
class Config:
    foo: int = 0
    bar: int = 0

def config(main: Command):
   return Config.from_file(main.config)

...

class CommandBacklog:
    config: Annoated[Config, Arg.destructured()]
```

and then manually merge `config` and `command_backlog.config` somewhere downstream.

**Reply by pawamoy** [reply]
> Well i'd firstly suggest https://github.com/DanCardin/dataclass-settings instead of dataclass-settings 😆

Nice!

> I think injecting from a Command instance is problematic because you're saying you need an instance of Command to already be instantiated in order for the command to then be updated with new overriding values.
Or i guess perhaps in at least this specific case, (in theory) a Command could be instantiated beforehand because you're only making use of it inside the context of a subcommand class. I do think the feasibility of this avenue very much depends on it being only accessible from a subcommand.

Good points. It indeed enables the logic to happen only for a subcommand :thinking:  But that's also the case, more generally, for dependencies, right? In a subcommand's `__call__` method, I cannot depend on a command that is not a parent command, right?

> My thought being that perhaps you could supply a parse (or a From is syntactic construct for a parse) which could "depend" on a command From(from_config) def from_config(main: Command):. Which would leave the specific method by which you pull out a specific field outside cappa since, I'm less certain I love this cappa.From(Command, "config.github_token").

Yes, `cappa.From(Command, "config.github_token")` is very limited, and specific to this current use-case only, so it's probably not a good idea to bake this "attribute access" logic into Cappa directly, and better to leave that to the user, through a callable and injected dependencies.

So, if we don't want to use partials, something like this maybe:

```python
def from_config(main: Command, attr_name: str):
    ...


option: Annotated[..., cappa.Arg(default=cappa.From(from_config, "my_option"))]
```

Basically, all args (starting at the second one) and kwargs passed to `cappa.From` are passed again to the callable (first arg), after dependencies have been resolved and injected.

> Arg.destructured()

I don't mind having all fields of the config optional (it's actually my case!), however destructuring the whole config into a subcommand is not something I'd want: there are many fields that aren't relevant to this particular subcommand. I wouldn't want them to appear as option in the CLI.

**Reply by DanCardin** [reply]
> Good points. It indeed enables the logic to happen only for a subcommand 🤔 But that's also the case, more generally, for dependencies, right? In a subcommand's __call__ method, I cannot depend on a command that is not a parent command, right?

I was thinking more about how you couldn't use this at the same level to apply to a sibling attribute.

> So, if we don't want to use partials, something like this maybe:

i just meant you'd probably write your own thing a la

```python
def from_config(attr: str):
    def wrapped(main: Command):
        ...
    return From(wrapped)
```

and `default=from_config('my_option')`. Not that splatting kwargs is off the table, just that it's not necessarily necessary.

> however destructuring the whole config into a subcommand is not something I'd want

Yea that's certainly a drawback. probably enough subclassing could get you the subset of commands that give you the api you want, but probably not (in pydantic) without affecting the end-shape of the config.

just a thought!

---

I think I can see how we may to be able to propagate parent command instance context down to subcommand during mapping for parse functions. at least getting the value there, technically less obvious at the moment. And i see no obvious long term API issues with it because it's context that should always exist and it matches the rest of the api sylistically.

I'm less certain about this default From thing. Not that it's bad, i just dont immediately see whether it's the right choice. You cant "just" use `parse` because you want it heirarchically last after attempting an env var lookup?

as i'm typing this, perhaps i'm convincing myself. I suppose you could implement `Env` in terms of a `From` and still only need to handle one thing. I'll have to see how it shakes out

**Reply by pawamoy** [reply]
> I was thinking more about how you couldn't use this at the same level to apply to a sibling attribute.

Oh yes, I would never expect this to work (requiring an instance of self to instantiate self??).

> i just meant you'd probably write your own thing a la

Yeah, closure or partial, fine by me if it makes it easier for Cappa :slightly_smiling_face: 

---

> I'm less certain about this default From thing. Not that it's bad, i just dont immediately see whether it's the right choice. You cant "just" use parse because you want it heirarchically last after attempting an env var lookup?

You're suggesting I try using `parse` instead, right? Can the main command be injected in the callable? Yeah I'd have to see how this all plays out regarding the order in which things are done, what is coerced again into the final type, etc. I'll report back :slightly_smiling_face:

**Reply by DanCardin** [reply]
not necessarily that you should use parse, just thinking through alternative apis. it's not going to work as-is because the upper commands are not something we inject at that point (currently).

but parse/mapping and default allocation both happen around the same places in the code, so both are presumably possible. `default` certainly has the more appropriate semantic if you're looking for it to fall back in the event it's not specified.

And ahh, i think i just grokked why you were talking about overriding the `|` operator. honestly probably makes sense.

**Reply by pawamoy** [reply]
> not necessarily that you should use parse, just thinking through alternative apis. it's not going to work as-is because the upper commands are not something we inject at that point (currently).

OK thanks for letting me know.The [docs](https://cappa.readthedocs.io/en/latest/arg.html#parse-dependencies) only mention `TypeView` as injectable indeed.

> but parse/mapping and default allocation both happen around the same places in the code, so both are presumably possible. default certainly has the more appropriate semantic if you're looking for it to fall back in the event it's not specified.

:tada: 

> And ahh, i think i just grokked why you were talking about overriding the | operator. honestly probably makes sense.

:smile: 

Yeah initially I showed an example with `default=[...]`, as "pick the first existing value in this list", but a default value can actually be a list I suppose (even though it's probably not a good idea to declare a mutable default at import time, instead of using a default factory). So then I thought about overriding `|`, so that we don't have to pass a list to `default` :slightly_smiling_face: Objects like `cappa.Env` can then support be OR'd by implementing some logic in `__or__` (basically, return first left-most existing value). It would probably require a wrapping structure to defer the logic to when Cappa is invoked.

```python
class Default:
    def __or__(self, other: Default | DefaultOr):
        return DefaultOr(self, *other.defaults if isinstance(other, DefaultOr) else other)

class DefaultOr:
    def __init__(self, *defaults: Default):
        self.defaults = defaults

    def __or__(self, other: Default | DefaultOr):
        return DefaultOr(*self, *other.defaults if isinstance(other, DefaultOr) else other)


class Env(Default): ...
```

Something like this, keeping a flat structure :slightly_smiling_face: Sorry, just having fun writing some code, I know you don't need my help to work out a solution :smile:

**Reply by DanCardin** [reply]
so i've encountered sort of an unfortunate fundamental flaw in this plan. in order to instantiate a `CommandMain` (for use in supplying a value to `github_token`, you **need** a `CommandBacklog` instance already. e.g. `CommandMain(config=...)` is invalid, you need `CommandMain(config=..., subcommand=CommandBacklog(...))`, which creates a loop.

At the point at which this mapping is happening, i **do** have the raw values that have been parsed, and/or the mapped values that have been mapped up to that point, but it's not so simple as supplying a fully instantiated upper-object instance, like it is during invoke-time.

Now, i've done this after implementing a decent amount of other refactors that I think make the defaulting system more robust (like this `From` object still makes **sense**, I'm just not sure how to supply it the value you seem to need), so i think some of the resulting changes are valuable, even if this strat is a dead end.

Thoughts?

**Reply by pawamoy** [reply]
I see, thank you for the explanation! That's unfortunate indeed :sweat_smile: It's OK, that's how the system is designed and I don't want you to start messing with it just for this use-case. I'm sure we can find another way. I'll try to find something and report back :slightly_smiling_face:

**Reply by pawamoy** [reply]
Ha, maybe `__post_init__` would do the job well? I'll try that. The main issue will still be that I somehow want the token to be required, while defaulting to specific values (env var, config value) :thinking: Not sure how this is usually achieved with plain dataclasses (in a context where Cappa isn't invoked). I'll run some experiments :smile:

**Reply by DanCardin** [reply]
Well like i said above, i think some of the changes i've made are potentially better. e.g. `default=Env("FOO") | Prompt('Gimme a value') | ValueFrom(get_config, 'foo')` I can still see how this would be a useful abstraction even if it doesn't solve your specific problem.

And it makes prompt/env/default handling a bit more normalized, which is good.

---

I wonder if, with a custom action, you could get this working.

```python
    config: An[
        Config,
        cappa.Arg(
            short="-c",
            long=True,
            parse=None,
            default=Config.from_default_location(),
            action=load_config,
        ]

def load_config(context: ParseContext, value: Value):
    config = Config.from_file(value.value)
    result['config'] = config
    result.setdefault('subcommand', {})['github_token'] = config.github_token
```

then all downstream cli args would naturally override the defaults you provide from the config. Although i guess this only works when the config file is supplied, not if it's defaulted... it also wouldn't be possible to make it aware of the specific subcommand being selected at this point in the cli parse....

It feels like you need some mutable shared state. like a global/thread-local that represents the config, so you can just access the already-loaded config directly from any callable

**Reply by DanCardin** [reply]
idk if this sounds dumb or not, but think similar-ish to click's `@pass_context` which is a lot more overloaded but does include at least provide a way to stash extra info.

```python
class StateData(TypeDict):  # optional, but seems like it'd be nice to keep type safety on what's ultimately a dict
    config: Config

class Config:
    @classmethod
    def from_file(cls, file: Path, state: State[StateData] | None = None):  # optional just so you could call it outside cappa
        result = cls(....)
        if state is not None:
            state.set(config=result)
        return result

    @classmethod
    def from_default_location(cls, file: Path, state: State[StateData] | None = None):
         return cls.from_file('foo.json', state=state)
...

    config: An[Config, cappa.Arg(long=True, parse=Config.from_file), default_factory=Config.from_default_location)]

...
    github_token: An[str, Arg(long=True, default=Env('GITHUB_TOKEN') | ValueFrom(from_config, 'github_token'))]

...

def from_config(state: State[StateData], attr: str):
    config = state.get('config')  # not sure how to type-safedly retrieve the key, if state.state is `StateData`, seems like you can't describe the type of a retrieval method so that it works?
    return config[attr]
```

e.g. in order:
* originally suggested default feature, minus a way to pass upper-level commands
* add State object that is just a user-accessible store of data
* Add Arg.default_factory as shorthand alias for `default=ValueFrom(default_factory)` (or not 🤷)
* make State injectable into parse/ValueFrom

**Reply by pawamoy** [reply]
1. The logic is reversed: we now "push" the config into subcommands instead of "pulling" it from parent commands. Why not!
2. I think that as long as I can inject something into a `default` callable, and the config has already been loaded when it's called, then yeah, this works :slightly_smiling_face: However I fail to see how that could work since you still have to instantiate the subcommand first, and then the main command, but maybe the default values are all resolved before any command is instantiated?

**Reply by DanCardin** [reply]
1. not sure if you're referring to my last comment or the "action" one. If action, then yes. if not then i think this is still pulling, but you do need to first record the state somewhere which i guess might be what you mean by "push".

2. per my last comment's though experiment, it would be recording the instantiated `Config` object into the `State` object standalone. So it avoids the problem of needing to instantiate the top-level dataclass object by recording a piece of data (in this case, one of your field's values) as a standalone piece of data that can be referenced by anything that occurs.

at the end of the parsing step, you're left with a nested dictionary of raw values `{config: 'file.json', 'subcommand': {'polar_token': 'foo', ...}}`. We iterate over the command fields, in order, and retrieve the value from the result. if you hit a subcommand, the sub-result is passed to a recursive call of the same function.

So this **does** rely on the subcommand being parsed last at any given level, but so long as that is true, it'd go:
1. `Command.map_result(CommandMain, {config: 'file.json', 'subcommand': {'polar_token': 'foo', ...}})`
2. call CommandMain.config's parse method on the raw 'file.json' value
   1. e.g. Config.from_file or Config.from_default_location
   2. into which State gets injected and you record the result onto
4. handle CommandMain.subcommand e.g. recurse
   1. parse CommandBacklog.github_token
   2. no value, evaluate default
   3. inject State into `from_config`, which now has the recorded `Config` instance
   4. read from it, return

**Reply by pawamoy** [reply]
I was indeed referring to 1) the custom action suggestion and 2) the state suggestion.

OK great, thanks for the explanation on the order things are processed in! Looks like a good solution to me! Are all the needed pieces already in place, or is there something new to implement?

**Reply by pawamoy** [reply]
Ah obviously we need to implement the `ValueFrom` thing, and the ability to OR it with `Env`.

**Reply by pawamoy** [reply]
So here's what I'm playing with. Since the state basically acts as a global variable, I moved it into a class variable (global too), in `CommandMain`. The class then provides a method to load and cache the config. A secondary function lets us retrieve an attribute from the config, and is to be used as an argument `default` callable.

```python
def from_config(attr_name: str) -> Any:
    config = CommandMain._load_config()  # should already be cached
    return getattr(config, attr_name)


@cappa.command(name="list")
@dataclass(kw_only=True)
class CommandBacklogList:
    """Command to list the backlog of issues."""

    github_token: An[
        str,
        cappa.Arg(
            short=False,
            long=True,
            default=cappa.Env("GITHUB_TOKEN") | cappa.FromValue(from_config, "github_token"),
        ),
        Doc("""A GitHub token. Recommended scopes: `repo`."""),
    ]


@cappa.command(name="insiders")
@dataclass(kw_only=True)
class CommandMain:
    """Command to manage your Insiders projects."""

    _CONFIG: ClassVar[Config | None] = None

    @staticmethod
    def _load_config(file: Path | None = None) -> Config:
        if CommandMain._CONFIG is None:
            CommandMain._CONFIG = Config.from_file(file) if file else Config.from_default_location()
        return CommandMain._CONFIG

    subcommand: An[
        CommandBacklogList,
        cappa.Subcommand(group=_GROUP_SUBCOMMANDS),
        Doc("The selected subcommand."),
    ]

    config: An[
        Config,
        cappa.Arg(
            short="-c",
            long=True,
            parse=_load_config,  # load and cache config!
            group=_GROUP_GLOBAL,
            propagate=True,
        ),
        Doc("Path to the configuration file."),
    ] = field(default_factory=_load_config)  # load and cache config!
```

Now the only thing missing is indeed support for `default=cappa.Env("GITHUB_TOKEN") | cappa.FromValue(from_config, "github_token"))` (whether `FromValue` is implemented in Cappa or not, which is irrelevant if users can implement it themselves).

**Reply by DanCardin** [reply]
It doesn't impl the concept of State yet, but https://github.com/DanCardin/cappa/pull/180 i believe minimally does the necessary bits for value retrieval/fallback.

Although relative to your above snippet it's called `ValueFrom` (and i haven't written any docs)

**Reply by pawamoy** [reply]
Fantastic! I'll try #180. Yeah `ValueFrom` sounds better, I reversed it by mistake.

**Reply by pawamoy** [reply]
Seems to work well! Two things:

- The field default factory is called even if I pass an argument, and it's called before the `parse` callable, meaning my default factory **must not** cache the config. I therefore changed it to `field(default_factory=Config.from_default_location)`. That also means the default config is always loaded at least once, and at maximum twice when no config path is specified (the first call to `from_config` will cache it). It's definitely not critical (not even worth a millisecond I'm sure).
- Looking at the code, I believe `ValueFrom` will only return the value if it's not `None`. Since `None` can be a valid config value, could it check instead if it's not `cappa.Empty` (I think you have such an object)? Or maybe accept a parameter in `ValueFrom` to provide a sentinel? I'm happy to modify my `from_config` function to transform my own `Unset` into your `Empty` so that it's correctly detected as "unset" (just like unset env vars). This is especially important for when another lazy default is present after it.

```python
def from_config(attr_name: str) -> Any:
    config = CommandMain._load_config()
    value = getattr(config, attr_name)
    return cappa.empty if value is config.unset else value
```

```python
def some_func() -> int | None:
    ...

def some_other_func() -> int:
    ...

cappa.Arg(
    # None is the sentinel here, all args after that are passed to some_func.
    # Not sure what's best here, maybe sentinel=None is better? but that "eats" this param name.
    # Maybe just require that the callable hardcodes cappa.Empty as sentinel.
    default=cappa.ValueFrom(some_func, None, 0, "a", other_arg="passed to some_func")
          | cappa.ValueFrom(some_other_func)
)
```

**Reply by pawamoy** [reply]
Or the callable could raise a Cappa exception instead of returning a sentinel:

```python
def from_config(attr_name: str) -> Any:
    config = CommandMain._load_config()
    value = getattr(config, attr_name)
    if value is config.unset:
        raise cappa.EmptyValue()
    return value
```

**Reply by DanCardin** [reply]
Yea I haven't yet lazified default factories, that should bring that down to 0 when not used. It was originally falling back on None because the only two sources of dynamic input were `Env` (where None` means it wasn't there) and Prompt/Confirm where None means there was no input.

And then further because in the context of a CLI command, you either have a value or not, and `None` would generally (always?) come from a lack of use of the cli arg.

So you're saying you have an explicitly set None config value with a further fallback that you dont want to default to?

**Reply by pawamoy** [reply]
> So you're saying you have an explicitly set None config value with a further fallback that you dont want to default to?

I don't, actually, but I thought it wouldn't be uncommon to have `None` as a valid default value. Other formats than TOML have null values, so it could be useful/required to distinguish between null and unset. Even not considering data formats, maybe we want to consider a set but empty environment variable to be `None`, not `""`.

```python
import os

def from_env(name: str) -> str | None:
    if value := os.getenv(name):
        return value
    return None  # ENVVAR= returns None, not ""

...

# This should stop at `None` and skip calling `something_else`.
cappa.Arg(default=cappa.ValueFrom(from_env, "VARNAME") | cappa.ValueFrom(something_else))
```

**Reply by DanCardin** [reply, marked as answer]
https://github.com/DanCardin/cappa/pull/180
